packages/completer-extension/src/index.ts

```diff
@@ -45,337 +41,187 @@ namespace CommandIDs {
-    app.commands.addCommand(CommandIDs.invokeConsole, {
-      execute: () => {
-        const id = consoles.currentWidget && consoles.currentWidget.id;
-
-        if (id) {
-          return app.commands.execute(CommandIDs.invoke, { id });
-        }
-      }
-    });
-
-    // Add console completer select command.
-    app.commands.addCommand(CommandIDs.selectConsole, {
-      execute: () => {
-        const id = consoles.currentWidget && consoles.currentWidget.id;
-
-        if (id) {
-          return app.commands.execute(CommandIDs.select, { id });
-        }
-      }
-    });
-
-    // Set enter key for console completer select command.
-    app.commands.addKeyBinding({
-      command: CommandIDs.selectConsole,
-      keys: ['Enter'],
-      selector: `.jp-ConsolePanel .jp-mod-completer-active`
-    });
+    serviceManager: ICompletionProviderManager
+  ): Promise<void> => {
+    serviceManager.registerProvider(new ContextCompleterProvider());
+    serviceManager.registerProvider(new KernelCompleterProvider());
   }
 };
 
-/**
- * An extension that registers notebooks for code completion.
- */
-const notebooks: JupyterFrontEndPlugin<void> = {
-  id: '@jupyterlab/completer-extension:notebooks',
-  requires: [ICompletionManager, INotebookTracker],
+const manager: JupyterFrontEndPlugin<ICompletionProviderManager> = {
+  id: COMPLETION_MANAGER_PLUGIN,
+  requires: [
+    INotebookTracker,
+    IEditorTracker,
+    IConsoleTracker,
+    ISettingRegistry
+  ],
+  provides: ICompletionProviderManager,
   autoStart: true,
   activate: (
     app: JupyterFrontEnd,
-    manager: ICompletionManager,
-    notebooks: INotebookTracker
-  ): void => {
-    // Create a handler for each notebook that is created.
-    notebooks.widgetAdded.connect((sender, panel) => {
-      const editor = panel.content.activeCell?.editor ?? null;
-      const session = panel.sessionContext.session;
-      // TODO: CompletionConnector assumes editor and session are not null
-      const connector = new CompletionConnector({ session, editor });
-      const handler = manager.register({ connector, editor, parent: panel });
-
-      const updateConnector = () => {
-        const editor = panel.content.activeCell?.editor ?? null;
-        const session = panel.sessionContext.session;
+    notebooks: INotebookTracker,
+    editorTracker: IEditorTracker,
+    consoles: IConsoleTracker,
+    settings: ISettingRegistry
+  ): ICompletionProviderManager => {
+    const AVAILABLE_PROVIDERS = 'availableProviders';
+    const PROVIDER_TIMEOUT = 'providerTimeout';
+    const manager = new CompletionProviderManager();
+    const updateSetting = (settingValues: ISettingRegistry.ISettings): void => {
+      const providersData = settingValues.get(AVAILABLE_PROVIDERS);
+      const timeout = settingValues.get(PROVIDER_TIMEOUT);
+      manager.setTimeout(timeout.composite as number);
+      const selectedProviders = providersData.user || providersData.composite;
```

Comment:
Yes, I think that most of the time, users only want to use one provider. By merging the result, users need to explicitly set the rank to negative to disable all other providers instead of just having the provider they want to use in the custom setting panel.